rkt/cp.go

```diff
@@ -0,0 +1,281 @@
+	cleanPath := filepath.Clean(path)
+
+	info, err := os.Stat(cleanPath)
+	if err != nil {
+		return cleanPath, fmt.Errorf("path not found")
+	}
+
+	if checkDir && !info.IsDir() {
+		return cleanPath, fmt.Errorf("path is not a directory")
+	}
+
+	return cleanPath, nil
+}
+
+// getAppNameForCp returns the app name to enter
+// If one was supplied in the flags then it's simply returned
+// If the PM contains a single app, that app's name is returned
+// If the PM has multiple apps, the names are printed and an error is returned
+func getAppNameForCp(p *pkgPod.Pod) (*types.ACName, error) {
+	if flagCpAppName != "" {
+		return types.NewACName(flagCpAppName)
+	}
+
+	// figure out the app name, or show a list if multiple are present
+	_, m, err := p.PodManifest()
+	if err != nil {
+		return nil, errwrap.Wrap(errors.New("error reading pod manifest"), err)
+	}
+
+	switch len(m.Apps) {
+	case 0:
+		return nil, fmt.Errorf("pod contains zero apps")
+	case 1:
+		return &m.Apps[0].Name, nil
+	default:
+	}
+
+	stderr.Print("pod contains multiple apps:")
+	for _, ra := range m.Apps {
+		stderr.Printf("\t%v", ra.Name)
+	}
+
+	return nil, fmt.Errorf("specify app using \"rkt cp --app= ...\"")
+}
+
+// copyFileOrDirectory copies a source file or directory to a destination directory.
+func copyFileOrDirectory(srcPath string, destDirPath string) error {
+	_, srcFilename := filepath.Split(srcPath)
+	destPath := filepath.Join(destDirPath, srcFilename)
+
+	return copyTree(srcPath, destPath)
+}
+
+// copyTree recursively copies a directory tree, preserving permissions.
+// Symlinks are ignored and skipped.
+func copyTree(src string, dst string) error {
+	si, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+
+	if si.Mode()&os.ModeSymlink != 0 {
+		return nil
+	}
+
+	if !si.IsDir() {
+		return copyFile(src, dst)
+	}
+
+	_, err = os.Stat(dst)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+
+		// Create the destination directory if it doesn't exist yet.
+		err = os.MkdirAll(dst, si.Mode())
+		if err != nil {
+			return err
+		}
+	}
+
+	entries, err := ioutil.ReadDir(src)
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range entries {
+		srcPath := filepath.Join(src, entry.Name())
+		dstPath := filepath.Join(dst, entry.Name())
+
+		err = copyTree(srcPath, dstPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// copyFile copies the contents of the file named src to the file named by dst.
+// The file will be created if it does not already exist.
+// If the no-clobber flag is set and the destination file exists, the destination file is not
+// changed. If the no-clobber flag is not set and the destination file exists, the content will be
+// replaced by the content of the source file. Permissions of the src file are applied to the dst
+// file.
+func copyFile(src, dst string) error {
+	if flagNoClobber {
+		if _, err := os.Stat(dst); err == nil {
+			return nil
+		}
+	}
+
+	in, err := os.Open(src)
```

Comment:
There's racyness present in the current implementation which will lead to this frequently erroring on a system with lots of activity, and could result in this copying the wrong thing.

The `Stat` and such done above could refer to a different file than ultimately gets opened here, so the symlink check could be false, the file could have been deleted, etc.

To make this not racy, a single fd will have to be preserved throughout the validate/copy/permissions steps.